src/main/java/io/confluent/connect/jdbc/sink/BufferedRecords.java

```diff
@@ -191,8 +191,9 @@ public BufferedRecords(
     log.trace("{} records:{} resulting in totalUpdateCount:{} totalDeleteCount:{}",
         config.insertMode, records.size(), totalUpdateCount, totalDeleteCount
     );
-    if (totalUpdateCount.filter(total -> total != expectedCount).isPresent()
-        && config.insertMode == INSERT) {
+    if (dbDialect.assertUpdateCount()
+            && totalUpdateCount.filter(total -> total != expectedCount).isPresent()
+            && config.insertMode == INSERT) {
```

Comment:
I think I agree with @gharris1727 that I would rather we move this logic into the dialect and use inheritance to have dialects that should skip this check override. What if we change the `assertUpdateCount` signature to take as input the `totalUpdateCount` and `expectedCount` and maybe insert mode and just called into the dialect here
```
dbDialect.assertUpdateCount(expectedCount, totalUpdateCount, INSERT);
```

Reply:
I can't think of any different ay to check this equality by dialect. I see only two variants compare or don't compare. In reality the comparison is kind of irrelevant if we check EXECUTE_FAILED in the loop and fail.

Reply:
If the comparison is going to be irrelevant, I don't see why we need to expose a method on the dialect just for it. Maybe we should just remove this logic in favor of catching this earlier in `executeUpdates`

Reply:
I would definitely prefer that if we all agree that there is no value from the actual reconcile

Reply:
I like this approach better. It keeps the check as is for the existing dialects while allowing new dialects to bypass this (which I assume will happen for a few drivers)

Reply:
The one thing we cold potentially do is to pass the `insertMode` into this and use it to expand this a bit further. and default impl would return true if config.insertMode == INSERT